ipn/ipnlocal/peerapi.go

```diff
@@ -341,11 +341,71 @@ func (s *peerAPIServer) DeleteFile(baseName string) error {
 // accidentally logging actual filenames anywhere.
 const redacted = "redacted"
 
+type redactedErr struct {
+	msg   string
+	inner error
+}
+
+func (re *redactedErr) Error() string {
+	return re.msg
+}
+
+func (re *redactedErr) Unwrap() error {
+	return re.inner
+}
+
 func redactErr(err error) error {
-	if pe, ok := err.(*os.PathError); ok {
+	var redactStrings []string
+
+	var pe *os.PathError
+	if errors.As(err, &pe) {
+		// If this is the root error, then we can redact it directly.
+		if err == pe {
+			pe.Path = redacted
+			return pe
+		}
+
+		// Otherwise, we have a *PathError somewhere in the error
+		// chain, and we can't redact it because something later in the
+		// chain may have cached the Error() return already (as
+		// fmt.Errorf does).
+		//
+		// Add this path to the set of paths that we will redact, below.
+		redactStrings = append(redactStrings, pe.Path)
+
+		// Also redact the Path value so that anything that calls
+		// Unwrap in the future gets the redacted value.
 		pe.Path = redacted
 	}
-	return err
+
+	var le *os.LinkError
+	if errors.As(err, &le) {
+		// If this is the root error, then we can redact it directly.
+		if err == le {
+			le.New = redacted
+			le.Old = redacted
+			return le
+		}
+
+		// As above
+		redactStrings = append(redactStrings, le.New, le.Old)
+		le.New = redacted
+		le.Old = redacted
+	}
+
+	if len(redactStrings) == 0 {
+		return err
+	}
+
+	s := err.Error()
```

Comment:
Hopefully `len(redactStrings)` isn't large, otherwise this is expensive.

Reply:
I wasn't able to find any case where we'd have more than two entries in the list, since I don't see any cases where an `*os.PathError` or `*os.LinkError` wrap either of those types. In the multi-error case, it's possible that this could get expensive with multiple distinct `*os.PathError`s wrapped into a single overarching error, at which point we would probably want to revisit this.